crates/dyn-abi/src/token.rs

```diff
@@ -134,29 +134,35 @@ impl<'a> DynToken<'a> {
         let dynamic = self.is_dynamic();
         match self {
             Self::Word(w) => *w = WordToken::decode_from(dec)?.0,
-            Self::FixedSeq(tokens, size) => {
+            Self::FixedSeq(_, _) => {
                 let mut child = if dynamic {
                     dec.take_indirection()?
                 } else {
                     dec.raw_child()
                 };
-                for token in tokens.to_mut().iter_mut().take(*size) {
-                    token.decode_populate(&mut child)?;
+
+                self.decode_sequence_populate(&mut child)?;
+
+                if !dynamic {
+                    dec.take_offset(child);
                 }
             }
             Self::DynSeq { contents, template } => {
                 let mut child = dec.take_indirection()?;
                 let size = child.take_u32()? as usize;
-                let mut new_tokens: Vec<DynToken<'a>> = Vec::with_capacity(size);
-                for _ in 0..size {
-                    let mut t = if let Some(item) = new_tokens.first() {
-                        item.clone()
-                    } else {
-                        *(template.take().unwrap())
-                    };
-                    t.decode_populate(&mut child)?;
-                    new_tokens.push(t);
-                }
+                // This appears to be an unclarity in the solidity spec. The
+                // spec specifies that offsets are relative to the beginning of
+                // `enc(X)`. But known-good test vectors have it relative to the
+                // word AFTER the array size
+                let mut child = child.raw_child();
+
+                let mut new_tokens: Vec<_> = Vec::with_capacity(size);
+                new_tokens.resize(size, *(template.take().unwrap()));
```

Comment:
this unwrap can panic, also unused parens

Reply:
if it panics, it's an implementation bug. I'll add an expect and a note why

the bug here is that the function should be pub(crate)